### PR TITLE
Propagating object params

### PR DESCRIPTION
**Pull request by chitrangpatel**
# Changes

Prior to this, object parameters were not being propagated down the spec during resolution.
After following the same logic applied for propagating parameters in PRs https://github.com/tektoncd/pipeline/pull/5291 and https://github.com/tektoncd/pipeline/pull/5143, we can not also propagate object params. This PR address issue https://github.com/tektoncd/pipeline/issues/5140

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
Propagating object params
```

### CODE REVIEW

**Comment by chitrangpatel** [comment]
@chuangw6 Please take a look when you can

**Comment by chitrangpatel** [comment]
I will work on getting the coverage to its original value/better tomorrow

File: pkg/apis/pipeline/v1/taskrun_validation.go [hunk 1]
```diff
@@ -116,15 +116,44 @@ func (ts *TaskRunSpec) validateInlineParameters(ctx context.Context) (errs *apis
...
+		for i, ps := range paramSpec {
 			if ps.Name == p.Name {
+				if p.Default != nil && p.Default.ObjectVal != nil {
+					for k, v := range p.Default.ObjectVal {
+						if paramSpec[i].Default.ObjectVal == nil {
+							paramSpec[i].Default.ObjectVal = map[string]string{k: v}
+						} else {
+							paramSpec[i].Default.ObjectVal[k] = v
+						}
+						if fmt.Sprintf("%T", v) == "string" {
```
Comment (lbernick) [inline pkg/apis/pipeline/v1/taskrun_validation.go:146]: Can you explain a bit what's going on here or maybe give a yaml example? this is if the literal value of an object param property is "string"? I don't understand why this is a special case or why we're using `sprintf` in this way

Comment (chitrangpatel) [reply pkg/apis/pipeline/v1/taskrun_validation.go:146]: Thanks for catching that @lbernick! I put it there originally thinking that the object params also allowed us to provide arrays as values. After going through the TEPs regarding object params yesterday, I figured out that we currently only support string values. I missed updating the conditional later. The tests passed because nothing really triggered a failure. I will fix it ASAP.

**Review by chuangw6 — COMMENTED** [review]
Thanks @chitrangpatel !

File: pkg/apis/pipeline/v1beta1/task_validation.go [hunk 2]
```diff
@@ -319,10 +319,6 @@ func (p ParamSpec) ValidateType() *apis.FieldError {
 // definition of `properties` section and the type of a PropertySpec is allowed.
 // (Currently, only string is allowed)
 func (p ParamSpec) ValidateObjectType() *apis.FieldError {
-	if p.Type == ParamTypeObject && p.Properties == nil {
-		return apis.ErrMissingField(fmt.Sprintf("%s.properties", p.Name))
-	}
-
```
[the remark this answers is not in the document]
Comment (chitrangpatel) [reply pkg/apis/pipeline/v1beta1/task_validation.go:325]: That's right. We cannot have this check because there might be no `properties` field defined when users are propagating parameters. Good catch though. When we explicitly declare the task without parameter propagation we want this check. I will fix accordingly and ping you later to verify.

File: pkg/apis/pipeline/v1beta1/task_validation_test.go [hunk 3]
```diff
@@ -119,6 +119,17 @@ func TestTaskSpecValidatePropagatedParamsAndWorkspaces(t *testing.T) {
...
+		name: "propagating object params valid step with script",
+		fields: fields{
+			Steps: []v1beta1.Step{{
+				Name:  "propagatingobjectparams",
+				Image: "my-image",
+				Script: `
+				#!/usr/bin/env bash
+				$(params.message.foo)`,
+			}},
+		},
```
[the remark this answers is not in the document]
Comment (chitrangpatel) [reply pkg/apis/pipeline/v1beta1/task_validation_test.go:132]: I can do that to avoid confusion. Thanks!

File: pkg/apis/pipeline/v1beta1/taskrun_validation_test.go [hunk 4]
```diff
@@ -177,6 +245,36 @@ func TestTaskRun_Validate(t *testing.T) {
...
+					Steps: []v1beta1.Step{{
+						Name:    "echo",
+						Image:   "ubuntu",
+						Command: []string{"echo"},
+						Args:    []string{"$(params.task-words.hello)"},
+					}},
+				},
+			},
+		},
+		wc: config.EnableAlphaAPIFields,
```
[the remark this answers is not in the document]
Comment (chitrangpatel) [reply pkg/apis/pipeline/v1beta1/taskrun_validation_test.go:277]: Will do!

File: pkg/apis/pipeline/v1beta1/taskrun_validation.go [hunk 5]
```diff
@@ -117,15 +117,44 @@ func (ts *TaskRunSpec) validateInlineParameters(ctx context.Context) (errs *apis
...
+		}
+		if p.Value.ObjectVal != nil {
+			prop := make(map[string]PropertySpec)
+			for k, v := range p.Value.ObjectVal {
+				if fmt.Sprintf("%T", v) == "string" {
+					prop[k] = PropertySpec{Type: ParamTypeString}
+				}
+			}
+			pSpec.Type = ParamTypeObject
+			pSpec.Properties = prop
```
[the remark this answers is not in the document]
Comment (chitrangpatel) [reply pkg/apis/pipeline/v1beta1/taskrun_validation.go:132]: @chuangw6 did you mean I replace `prop[k] = PropertySpec{Type: ParamTypeString}` with `prop[k] = PropertySpec{Type: p.Value.Type}`?

**Comment by chitrangpatel** [comment]
@lbernick @chuangw6 feel free to review again. I have incorporated your changes.

File: pkg/apis/pipeline/v1/taskrun_validation_test.go [hunk 6]
```diff
@@ -111,6 +131,28 @@ func TestTaskRun_Validate(t *testing.T) {
 			},
 		},
 		wc: config.EnableAlphaAPIFields,
+	}, {
+		name: "propagating partial params with different provided and default names",
```
Comment (lbernick) [inline pkg/apis/pipeline/v1/taskrun_validation_test.go:135]: this test case has the same name as the next one, can you make one or both more descriptive?

File: pkg/apis/pipeline/v1/taskrun_validation_test.go [hunk 7]
```diff
@@ -142,6 +184,32 @@ func TestTaskRun_Validate(t *testing.T) {
 			},
 		},
 		wc: config.EnableAlphaAPIFields,
+	}, {
+		name: "propagating partial object params with different provided and default names",
```
Comment (lbernick) [inline pkg/apis/pipeline/v1/taskrun_validation_test.go:188]: I think I left a similar comment on a previous PR but params don't have default names (just default values), so I think this is a bit confusing-- it looks like here the task spec is declaring another param task-words-2 and not using it (which I'm sort of surprised is valid); the test case name should reflect that. Also what's a "partial" param?

File: pkg/apis/pipeline/v1/taskrun_validation_test.go [hunk 8]
```diff
@@ -176,6 +244,36 @@ func TestTaskRun_Validate(t *testing.T) {
 			},
 		},
 		wc: config.EnableAlphaAPIFields,
+	}, {
+		name: "propagating partial object params in taskrun",
```
Comment (lbernick) [inline pkg/apis/pipeline/v1/taskrun_validation_test.go:248]:
```suggestion
		name: "propagating object params with multiple keys",
```

File: pkg/apis/pipeline/v1/taskrun_validation_test.go [hunk 9]
```diff
@@ -202,6 +300,32 @@ func TestTaskRun_Validate(t *testing.T) {
 			},
 		},
 		wc: config.EnableAlphaAPIFields,
+	}, {
+		name: "propagating object params with taskrun same names",
```
Comment (lbernick) [inline pkg/apis/pipeline/v1/taskrun_validation_test.go:304]: this is technically not propagated params right? Just normal params.

File: pkg/apis/pipeline/v1/taskrun_validation_test.go [hunk 10]
```diff
@@ -176,6 +244,36 @@ func TestTaskRun_Validate(t *testing.T) {
...
+						Default: &v1.ParamValue{
+							Type:      v1.ParamTypeObject,
+							ObjectVal: map[string]string{"hello": "task run def", "hi": "task"},
+						},
+					}},
+					Steps: []v1.Step{{
+						Name:    "echo",
+						Image:   "ubuntu",
+						Command: []string{"echo"},
+						Args:    []string{"$(params.task-words.hello)", "$(params.task-words.hi)"},
```
Comment (lbernick) [inline pkg/apis/pipeline/v1/taskrun_validation_test.go:271]: this is interesting-- it seems like the task spec has defined a default value for the object param which contains the key "hi", and the default value is overridden by the value defined in the taskrun, which doesn't contain the key "hi". Why is it valid to use the key "hi" in the args? When a taskrun overrides the default value for an object param, is that on a per-key basis or for the param as a whole? Is this documented anywhere?

File: pkg/apis/pipeline/v1/taskrun_validation_test.go (hunk 7 above)
Comment (chitrangpatel) [reply pkg/apis/pipeline/v1/taskrun_validation_test.go:188]: Yes, I had that for increasing coverage. I will use it to avoid confusion. Partial params in my mind (which you appropriately worded in the next comment below) is really params where some of the keys are provided by the task run/pipelinerun and others are defaulted in the task/pipeline spec. (i.e. a case where only some of the params are explicitly provided).

File: pkg/apis/pipeline/v1/taskrun_validation_test.go (hunk 6 above)
Comment (chitrangpatel) [reply pkg/apis/pipeline/v1/taskrun_validation_test.go:135]: Sure.

File: pkg/apis/pipeline/v1/taskrun_validation_test.go (hunk 10 above)
Comment (chitrangpatel) [reply pkg/apis/pipeline/v1/taskrun_validation_test.go:271]: That's a good question. Below is an example that should explain a few things. In the original [TEP](https://github.com/tektoncd/community/blob/main/teps/0107-propagating-parameters.md) on propagating params, we captured the usage for string and array types. But the main idea was that we wanted to have a [scoping mechanism](https://github.com/tektoncd/community/blob/main/teps/0107-propagating-parameters.md#scope-and-precedence). In this case, the keys provided at the task run will replace the default keys during substitution.  The scoping part is documented in the context of string and array types. Thanks for bringing this up. It seems like we should also update the documentation to show object replacements. 

```yaml
apiVersion: tekton.dev/v1beta1
kind: TaskRun
metadata:
...
spec:
  params:
  - name: gitrepo
    value:
      commit: sha123
      url: xyz.com
  serviceAccountName: default
  taskSpec:
    params:
    - default:
        branch: main
        commit: sha456
        url: abc.com
      name: gitrepo
      type: object
    steps:
    - args:
      - echo
      - --url=$(params.gitrepo.url)
      - --commit=$(params.gitrepo.commit)
      - --branch=$(params.gitrepo.branch)
...
status:
...
  taskSpec:
    params:
    - default:
        branch: main
        commit: sha456
        url: abc.com
      name: gitrepo
      type: object
    steps:
    - args:
      - echo
      - --url=xyz.com
      - --commit=sha123
      - --branch=main
...
```

File: pkg/apis/pipeline/v1beta1/taskrun_validation.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (chuangw6) [reply pkg/apis/pipeline/v1beta1/taskrun_validation.go]: I meant to say `pSpec.Type = p.value.Type` to make the type setting easier & more accurate for the fake spec, and we should keep `prop[k] = PropertySpec{Type: ParamTypeString}`. 

Chatted with @chitrangpatel offline. we aligned on this.

File: pkg/apis/pipeline/v1/taskrun_validation_test.go (hunk 10 above)
Comment (lbernick) [reply pkg/apis/pipeline/v1/taskrun_validation_test.go:305]: I think this is actually more related to TEP-0075 than TEP-0107. In TEP-0075, under [declaring object results and params](https://github.com/tektoncd/community/blob/main/teps/0075-object-param-and-result-types.md#declaring-object-results-and-params), it states "there will be no behavior to merge the default with the provided value, if for example the provided value specified some fields but not others". This sounds like if a user wants to replace the default value of an object param, any keys they do not specify will not have values, unlike the status.taskSpec.steps in your example.

@chuangw6 do you know if this behavior is validated anywhere? Would one of you be able to add documentation for it? I think it would also be reasonable to support the behavior shown in this example, but we would need to update TEP-0075 first.

Comment (chuangw6) [reply pkg/apis/pipeline/v1/taskrun_validation_test.go:305]: Yes, this is validated in [webhook](https://github.com/tektoncd/pipeline/blob/af4d0e6eb018a34c950180939a7f6a2f24941da6/pkg/apis/pipeline/v1beta1/task_validation.go#L408-L426). Since TEP75 proposes that `properties` section is required for either inline or ref definition of the spec, the validation makes sure the key used in steps is indeed declared in `properties`. The way I interpreted this line *"there will be no behavior to merge the default with the provided value, if for example the provided value specified some fields but not others"* in the TEP75 was that if an object `default` is provided, the value for all keys must be provided. Same for the value provided in taskrun. 

When it comes to propagating object params, I agree that it makes sense and provides more flexibility to propagate on   a per-key basis like the example @chitrangpatel shows above. @chitrangpatel also proposed that `properties` section is not required when it comes to propagating object param (like the example above), which I think makes sense too. This is because param propagating only happens for inline spec. As long keys being referenced exist in either `default` or `value`, it would be valid usage. (@chitrangpatel Please correct me if I'm wrong.)

So now it's about updating the documentation like TEP75 and [tasks.md](https://github.com/tektoncd/pipeline/blob/main/docs/tasks.md#parameter-type):
- the merge behaviour between `default` and `value` is allowed. And `value` from taskruns takes precedence. (FYI: #5270 might be related too.)
- `properties` section is not required when the spec is inline (embedded).

Did I miss anything here?

File: pkg/apis/pipeline/v1beta1/taskrun_validation.go [hunk 11]
```diff
@@ -120,17 +120,49 @@ func (ts *TaskRunSpec) validateInlineParameters(ctx context.Context) (errs *apis
...
+						}
+						if paramSpec[i].Properties == nil {
+							paramSpec[i].Properties = map[string]PropertySpec{
+								k: {Type: ParamTypeString},
+							}
+						} else {
+							paramSpec[i].Properties[k] = PropertySpec{Type: ParamTypeString}
+						}
+					}
+				}
```
Comment (chuangw6) [inline pkg/apis/pipeline/v1beta1/taskrun_validation.go:153]: I am wondering if we can change `paramSpec` from `[]ParamSpec` to `map[string]ParamSpec` type (paramname:paramSpec) to avoid the 3-layer for loops. When all ParamSpecs are needed at the end for validation, we can just use a single for loop to get all the values of the map. wdyt?

nit: maybe also rename `paramSpec` to `paramSpecsForValidation` or just add a comment to explain what is this variable for?

File: pkg/apis/pipeline/v1/taskrun_validation_test.go (hunk 10 above)
Comment (lbernick) [reply pkg/apis/pipeline/v1/taskrun_validation_test.go:305]: Could you point me to any test cases that cover this?

I agree that having defaults overridden on a per-key basis is much more useful behavior, but that's not how I interpreted TEP-0075. It does look like #5270 is the feature we're describing. Can we update TEP-0075 as you've suggested?

No strong opinions about whether `properties` should be required in param propagation, but that should likely also be covered in TEP-0107.

Comment (chuangw6) [reply pkg/apis/pipeline/v1/taskrun_validation_test.go:305]: > Could you point me to any test cases that cover this?

Sure [here](https://github.com/tektoncd/pipeline/blob/af4d0e6eb018a34c950180939a7f6a2f24941da6/pkg/apis/pipeline/v1beta1/pipeline_validation_test.go#L1989-L2132) is the tests that cover this. I realized that the test for covering task steps was missed. I just added that https://github.com/tektoncd/pipeline/pull/5411. PTAL. Thanks @lbernick 


> I agree that having defaults overridden on a per-key basis is much more useful behavior, but that's not how I interpreted TEP-0075. It does look like https://github.com/tektoncd/pipeline/issues/5270 is the feature we're describing. Can we update TEP-0075 as you've suggested?

Sure. I'll update the TEP.

File: pkg/apis/pipeline/v1beta1/taskrun_validation.go (hunk 11 above)
Comment (chitrangpatel) [reply pkg/apis/pipeline/v1beta1/taskrun_validation.go:153]: I can change it to a map, however, it will require changing many other calls too since `[]ParamSpec` is required by `ValidatePipelineParams` and many other functions. They will all need to be changed. Might be outside the scope of just this PR.

Comment (chitrangpatel) [reply pkg/apis/pipeline/v1beta1/taskrun_validation.go:153]: I can certainly change `paramSpec` to `paramSpecForValidation`.

Comment (chuangw6) [reply pkg/apis/pipeline/v1beta1/taskrun_validation.go:153]: In this function, we just called `ValidateParameterVariables` at the end to delegate the validation job. Since it expects a slice of paramspec for the third argument, we can just collect the map's values using a single loop right before calling this func. This may help us avoid layered loops?

Comment (chitrangpatel) [reply pkg/apis/pipeline/v1beta1/taskrun_validation.go:153]: I see what you mean. So basically just convert the map to a slice of param spec at the end before passing it for validation.

Comment (chuangw6) [reply pkg/apis/pipeline/v1beta1/taskrun_validation.go:153]: Yes!

Comment (chitrangpatel) [reply pkg/apis/pipeline/v1beta1/taskrun_validation.go]: Done!

File: pkg/apis/pipeline/v1/taskrun_validation_test.go (hunk 7 above)
Comment (chitrangpatel) [reply pkg/apis/pipeline/v1/taskrun_validation_test.go]: I updated the names. Should be ok now.

File: pkg/apis/pipeline/v1beta1/taskrun_validation.go [hunk 12]
```diff
@@ -114,27 +114,64 @@ func (ts *TaskRunSpec) validateInlineParameters(ctx context.Context) (errs *apis
...
 			}
+		} else {
+			if p.Default != nil && p.Default.ObjectVal != nil {
+				prop := make(map[string]PropertySpec)
+				for k := range p.Default.ObjectVal {
+					prop[k] = PropertySpec{Type: ParamTypeString}
+				}
+				p.Properties = prop
+			}
+			paramSpecForValidation[p.Name] = p
```
Comment (chuangw6) [inline pkg/apis/pipeline/v1beta1/taskrun_validation.go:166]: Feel like this part is a bit hard to read. How about decomposing them into helper function that does the merge job? 

Two benefits I can think of: 1. easy to understand, 2. it can be reused in `pipelinerun_validation.go` as well.
```
=========
                var the_map map[string]string
		if v, ok := paramSpecForValidation[p.Name]; ok {
			the_map = mergeObjectPartOfParamValue(v.Default, p.Default)
		} else {
			the_map = mergeParamValueObjectPart(nil, p.Default)
		}
		paramSpecForValidation[p.Name].Default = the_map
		paramSpecForValidation[p.Name].Properties = composePropertySectionFromAMap(the_map)
		
==========	
helpers
==========
		
func mergeObjectPartOfParamValue(input1 ParamValue, input2 ParamValue) map[string]string {
             // check nil of input1 and input2
             // if same key exist in both, value in input1 will take precedence 
}
		
func composePropertySectionFromAMap(input map[string]string) map[string]PropertySpec {
}
```

File: pkg/apis/pipeline/v1beta1/taskrun_validation_test.go [hunk 13]
```diff
@@ -177,6 +284,37 @@ func TestTaskRun_Validate(t *testing.T) {
...
+					Steps: []v1beta1.Step{{
+						Name:    "echo",
+						Image:   "ubuntu",
+						Command: []string{"echo"},
+						Args:    []string{"$(params.task-words.hello)", "$(params.task-words.hi)"},
+					}},
+				},
+			},
+		},
+		wc: config.EnableAlphaAPIFields,
```
Comment (chuangw6) [inline pkg/apis/pipeline/v1beta1/taskrun_validation_test.go:317]: Without `properties`, this tricks the [check](https://github.com/tektoncd/pipeline/blob/0fe293118eccdbf4dd478f00ef8d41e0bff06af5/pkg/reconciler/taskrun/validate_resources.go#L86-L88) on the reconciler side to pass because the check compares what are provided on the taskrun level against what are needed in the spec's `properties` section. And the needed is empty so the check is passed.

Although taskrun will not need to provide values for all keys required in `properties` because of https://github.com/tektoncd/pipeline/issues/5270 , the reconciler side will still need to compare things (combined key set provided in taskrun and spec's default) against `properties` section anyway.

File: pkg/apis/pipeline/v1beta1/taskrun_validation_test.go [hunk 14]
```diff
@@ -112,6 +133,59 @@ func TestTaskRun_Validate(t *testing.T) {
...
+					Steps: []v1beta1.Step{{
+						Name:    "echo",
+						Image:   "ubuntu",
+						Command: []string{"echo"},
+						Args:    []string{"$(params.task-words.hello)"},
+					}},
+				},
+			},
+		},
+		wc: config.EnableAlphaAPIFields,
```
Comment (chuangw6) [inline pkg/apis/pipeline/v1beta1/taskrun_validation_test.go:188]: I am a bit doubt if this a real case for propagating param. If no value is provided in taskrun, why does user write something about that param in taskrun?

File: pkg/apis/pipeline/v1beta1/taskrun_validation.go (hunk 12 above)
Comment (chitrangpatel) [reply pkg/apis/pipeline/v1beta1/taskrun_validation.go]: Agreed. I was about to do that after I got all the tests passing.

File: pkg/apis/pipeline/v1beta1/task_validation.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (chitrangpatel) [reply pkg/apis/pipeline/v1beta1/task_validation.go]: @chuangw6, based on our discussion,  I updated this line to:
```
if p.Type == ParamTypeObject && p.Default != nil && p.Properties == nil {
```
This will enforce that if the defaults are specified then properties must also be provided.

File: pkg/apis/pipeline/v1beta1/task_validation.go [hunk 15]
```diff
@@ -366,10 +362,10 @@ func ValidateParameterVariables(ctx context.Context, steps []Step, params []Para
 	errs = errs.Also(validateNameFormat(stringParameterNames.Insert(arrayParameterNames.List()...), objectParamSpecs))
 	if config.ValidateParameterVariablesAndWorkspaces(ctx) == true {
 		errs = errs.Also(validateVariables(ctx, steps, "params", allParameterNames))
+		errs = errs.Also(validateObjectDefault(objectParamSpecs))
+		errs = errs.Also(validateObjectUsage(ctx, steps, objectParamSpecs))
 	}
-	errs = errs.Also(validateArrayUsage(steps, "params", arrayParameterNames))
-	errs = errs.Also(validateObjectDefault(objectParamSpecs))
-	return errs.Also(validateObjectUsage(ctx, steps, objectParamSpecs))
+	return errs.Also(validateArrayUsage(steps, "params", arrayParameterNames))
```
[the remark this answers is not in the document]
Comment (chitrangpatel) [reply pkg/apis/pipeline/v1beta1/task_validation.go:372]: I'm not sure. The tests were passing without including it so I didn't want to change the default behaviour.

File: pkg/apis/pipeline/v1beta1/taskrun_validation_test.go (hunk 13 above)
Comment (chitrangpatel) [reply pkg/apis/pipeline/v1beta1/taskrun_validation_test.go]: I now provide properties after our discussion today.

Comment (chitrangpatel) [reply pkg/apis/pipeline/v1beta1/taskrun_validation_test.go]: Resolve this if you are now satisfied.

File: pkg/apis/pipeline/v1beta1/taskrun_validation_test.go (hunk 14 above)
Comment (chitrangpatel) [reply pkg/apis/pipeline/v1beta1/taskrun_validation_test.go]: I had this in place to increase coverage. Because of the behaviour of propagated params, when we combine defaults with provided values, I need to cover different cases.

File: pkg/apis/pipeline/v1beta1/taskrun_validation.go (hunk 12 above)
Comment (chitrangpatel) [reply pkg/apis/pipeline/v1beta1/taskrun_validation.go]: Done now.

File: pkg/apis/pipeline/v1beta1/pipelinerun_validation_test.go [hunk 16]
```diff
@@ -360,6 +544,7 @@ func TestPipelineRun_Validate(t *testing.T) {
 						Params: []v1beta1.Param{{
 							Name: "task-words",
 							Value: v1beta1.ParamValue{
+								Type:     v1beta1.ParamTypeArray,
```
Comment (jerop) [inline pkg/apis/pipeline/v1beta1/pipelinerun_validation_test.go:547]: why is type array added in some of these test cases? it doesn't seem related to the changes in this commit -- would rather not add them in this commit if not necessary

**Review by jerop — COMMENTED** [review]
could you please examples with propagated object params to the docs? existing examples only show propagated string params and it'd be helpful to make it clear that propagated object params are supported too

File: pkg/apis/pipeline/v1beta1/pipelinerun_validation_test.go (hunk 16 above)
Comment (chitrangpatel) [reply pkg/apis/pipeline/v1beta1/pipelinerun_validation_test.go:547]: That's because Chuang pointed out that we should be ParamType checking as well when we are propagating parameters. I had overlooked it and so my original test cases didn't have that. When I added the Type checking in line 189 of `pipelinerun_validation.go` to do it properly for object params, the arrays tests failed as well. Hence this change.

File: pkg/apis/pipeline/v1beta1/pipelinerun_validation.go [hunk 17]
```diff
@@ -156,66 +156,62 @@ func (ps *PipelineRunSpec) validateInlineParameters(ctx context.Context) (errs *
...
 		}
 	}
+	var paramSpec []ParamSpec
+	for _, v := range paramSpecForValidation {
+		paramSpec = append(paramSpec, v)
+	}
 	if ps.PipelineSpec != nil && ps.PipelineSpec.Tasks != nil {
 		for _, pt := range ps.PipelineSpec.Tasks {
 			if pt.TaskSpec != nil && pt.TaskSpec.Steps != nil {
+				errs = errs.Also(ValidateParameterTypes(ctx, paramSpec))
```
Comment (chitrangpatel) [inline pkg/apis/pipeline/v1beta1/pipelinerun_validation.go:189]: @jerop this is the line that caused the array test to fail and so I needed to add the Type field.

File: pkg/apis/pipeline/v1beta1/pipelinerun_validation_test.go (hunk 16 above)
Comment (chitrangpatel) [reply pkg/apis/pipeline/v1beta1/pipelinerun_validation_test.go:547]: Let me know.

File: pkg/apis/pipeline/v1/taskrun_validation_test.go (hunk 10 above)
Comment (lbernick) [reply pkg/apis/pipeline/v1/taskrun_validation_test.go:356]: @chitrangpatel I think this test case should result in an error until we resolve https://github.com/tektoncd/pipeline/pull/5427

Comment (chitrangpatel) [reply pkg/apis/pipeline/v1/taskrun_validation_test.go:356]: In that case, can we put this PR on hold until https://github.com/tektoncd/pipeline/pull/5427 is resolved? A lot of parameter propagation validation is around combining provided values with the defaults and comparing against the Properties field.

Comment (lbernick) [reply pkg/apis/pipeline/v1/taskrun_validation_test.go:356]: sure, fine with me

**Review by vdemeester — APPROVED** [review]

**Comment by chitrangpatel** [comment]
The following is the coverage report on the affected files.
Say `/test chitrang-test-1` to re-run this coverage report

File | Old Coverage | New Coverage | Delta
---- |:------------:|:------------:|:-----:
[pkg/apis/pipeline/v1/pipeline_validation.go](https://storage.cloud.google.com/tekton-prow/pr-logs/pull/tektoncd_pipeline/5371/chitrang-test-1/51421370456/artifacts/line-cov.html#file0) | 99.4% | 98.4% | -1.1
[pkg/apis/pipeline/v1/task_validation.go](https://storage.cloud.google.com/tekton-prow/pr-logs/pull/tektoncd_pipeline/5371/chitrang-test-1/51421370456/artifacts/line-cov.html#file1) | 97.0% | 97.4% | 0.4
[pkg/apis/pipeline/v1/taskrun_validation.go](https://storage.cloud.google.com/tekton-prow/pr-logs/pull/tektoncd_pipeline/5371/chitrang-test-1/51421370456/artifacts/line-cov.html#file2) | 98.4% | 99.2% | 0.8
[pkg/apis/pipeline/v1beta1/pipeline_validation.go](https://storage.cloud.google.com/tekton-prow/pr-logs/pull/tektoncd_pipeline/5371/chitrang-test-1/51421370456/artifacts/line-cov.html#file3) | 99.6% | 100.0% | 0.4
[pkg/apis/pipeline/v1beta1/pipelinerun_validation.go](https://storage.cloud.google.com/tekton-prow/pr-logs/pull/tektoncd_pipeline/5371/chitrang-test-1/51421370456/artifacts/line-cov.html#file4) | 97.4% | 98.5% | 1.1
[pkg/apis/pipeline/v1beta1/task_validation.go](https://storage.cloud.google.com/tekton-prow/pr-logs/pull/tektoncd_pipeline/5371/chitrang-test-1/51421370456/artifacts/line-cov.html#file5) | 97.1% | 97.6% | 0.4
[pkg/apis/pipeline/v1beta1/taskrun_validation.go](https://storage.cloud.google.com/tekton-prow/pr-logs/pull/tektoncd_pipeline/5371/chitrang-test-1/51421370456/artifacts/line-cov.html#file6) | 98.4% | 99.2% | 0.8
